keps/sig-auth/NNNN-20201209-client-exec-proxy/README.md

```diff
@@ -0,0 +1,711 @@
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+
+The motivation for this KEP stems from the inability of the existing authentication mechanisms to satisfy the author's requirement to implement a Kubernetes client that implements AWS's request signing process, known as Signature Version 4 (sig-v4).  In order to implement sig-v4, a canonical version of the request must be created, combined with any requered additional metadata, signed by the secret key, and the signature appended to the request in a header or query string parameter.  In order to achieve this, the request signer must have access to the entire request, so the existing exec-based authentication is unsatisfactory.  A generic solution is desirable to satisfy all other future request signing feature requests.
+
+The solution that exists today is to use the existing proxy configuration in client-go.  The includes configuring a proxy via the environment variables, HTTP_PROXY or HTTPS_PROXY, or using the explicit proxy URL supported by kubeconfig [[3]](https://github.com/kubernetes/kubernetes/pull/81443).  The problem with the solutions that exists today is that the operator experience of setting up and managing such a solution is poor.  For example, to implement a request signing proxy for kubelet, one has to manage the proxy process with an init system of some kind, and secure the proxy endpoint, both of which require significant, non-trivial configuration outside of Kubernetes.
+
+### Goals
+
+<!--
+List the specific goals of the KEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+The goals of this KEP include:
+1. Defining a specification which enables Kubernetes clients to do arbitrary request modification before sending the request to the API server, via an extensible mechanism to support the following use cases:
+    - Client request signing.
+    - Using hardware protected keys for request signing or mTLS.
+    - Adding or manipulating arbitrary headers to the request.
+2. Defining a specification which would provide the ability for the client-side proxy server to be executed and communicated with according to configuration solely contained in a kubeconfig file.
+3. Defining a specification which is interoperable with an existing client-side proxy setup, e.g. a proxy configured with the HTTP_PROXY environment variable.
+4. Defining a specification for a client-side proxy server that is easy to secure without taking additional setup steps, by:
+    - Using a unix domain socket for the communication between the client and the proxy, secured with file permissions.
```

Comment:
Please remove the UDS bits.  That is a separate feature KEP that is orthogonal to this one.